### PR TITLE
Dataset class with user defined label

### PR DESCRIPTION
**Pull request by jhgoh**
미리 여러 데이터셋을 섞고 weight를 계산해 데이터에 저장 후 train, validation, test set으로 구분하는 기존의 방법에서 비효율성이 있습니다.
이 PR에서는 데이터 파일 내에 label이나 weight 자체를 저장하는 대신 user가 직접 training 스크립트에서 지정할 수 있게 합니다. 이는 TMVA와 약간 닮은 면이 있습니다.

예를 들어 다음 코드와 같은 식으로 해서 샘플을 dataset에 추가하고, 동시에 (cross section)/(nGenEvent) 를 세팅해 normalize해줍니다.
Process 이름에 따라 어떤 번호로 label을 줄지 지정할 수 있습니다. Multiclassification을 할 경우에 대비해 addSignal 이나 addBackground등의 함수 형태를 쓰지 않았습니다.

```
myDataset = MyDataset()
myDataset.addSample("RPV_1400", "../data/CMS2018_unmerged/RPV/Gluino1400GeV/*/*.h5", weight=0.013/330599)
#myDataset.addSample("QCD_HT700to1000" , "../data/CMS2018_unmerged/QCD/HT700to1000/*/*.h5", weight=???)
#myDataset.addSample("QCD_HT1000to1500", "../data/CMS2018_unmerged/QCD/HT1000to1500/*/*.h5", weight=1094./15466225)
myDataset.addSample("QCD_HT1500to2000", "../data/CMS2018_unmerged/QCD/HT1500to2000/*/*.h5", weight=99.16/3199737)
myDataset.addSample("QCD_HT2000toInf" , "../data/CMS2018_unmerged/QCD/HT2000toInf/*/*.h5", weight=20.25/1520178)
myDataset.setProcessLabel("RPV_1400", 1)
myDataset.setProcessLabel("QCD_HT1500to2000", 0) ## This is not necessary because the default is 0
myDataset.setProcessLabel("QCD_HT2000toInf", 0) ## This is not necessary because the default is 0
myDataset.initialize(nWorkers=args.nreader)

lengths = [int(0.6*len(myDataset)), int(0.2*len(myDataset))]
lengths.append(len(myDataset)-sum(lengths))
trnDataset, valDataset, testDataset = torch.utils.data.random_split(myDataset, lengths)

```

위의 initialize 함수에서 이미지에 대한 텐서로의 변환이나, 전체 rescaling을 위한 sumWeight 계산 등이 이루어집니다.

Training, validation, testing set의 분리는 torch.utils.data.random_split을 이용합니다.

sample imbalance문제 해결을 위한 rescaling은 실제 training과 validation 단계에서 weight에 rescale factor를 직접 곱해주는 방식으로 하게 되는데, 이것은 같은 dataloader를 이용해 test set 그림을 그릴 수 있다고 판단했기 때문입니다.

Related issue: #71

### CODE REVIEW

**Comment by jhgoh** [comment]
Bad news: does not go up beyond 88% in some reason... to be understood.

**Comment by jhgoh** [comment]
Talking about the speed:
- 20 iteration x128 batch / second

**Comment by jhgoh** [comment]
#71

**Comment by jhgoh** [comment]
수행속도:
```epoch 1/50: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1609/1609 [01:19<00:00, 20.33it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 269/269 [00:09<00:00, 28.51it/s]
epoch 2/50: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1609/1609 [01:15<00:00, 21.34it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 269/269 [00:09<00:00, 28.61it/s]
epoch 3/50: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1609/1609 [01:15<00:00, 21.23it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 269/269 [00:09<00:00, 28.09it/s]
```

**Comment by jhgoh** [comment]
리소스 사용 모니터링 
image

**Comment by jhgoh** [comment]
현재 테스트로 사용중인 파일들과 구조

```
tree . -h
.
├── [4.0K]  QCD
│   ├── [4.0K]  HT1500to2000
│   │   └── [4.0K]  0000
│   │       ├── [ 34M]  hepcnn_10.h5
│   │       ├── [ 34M]  hepcnn_11.h5
│   │       ├── [ 34M]  hepcnn_12.h5
│   │       ├── [ 34M]  hepcnn_13.h5
│   │       ├── [ 34M]  hepcnn_14.h5
│   │       ├── [ 34M]  hepcnn_15.h5
│   │       ├── [ 34M]  hepcnn_16.h5
│   │       ├── [ 34M]  hepcnn_17.h5
│   │       ├── [ 34M]  hepcnn_18.h5
│   │       ├── [ 34M]  hepcnn_19.h5
│   │       ├── [ 34M]  hepcnn_1.h5
│   │       ├── [ 34M]  hepcnn_2.h5
│   │       ├── [ 34M]  hepcnn_3.h5
│   │       ├── [ 34M]  hepcnn_4.h5
│   │       ├── [ 34M]  hepcnn_5.h5
│   │       ├── [ 34M]  hepcnn_6.h5
│   │       ├── [ 34M]  hepcnn_7.h5
│   │       ├── [ 34M]  hepcnn_8.h5
│   │       └── [ 34M]  hepcnn_9.h5
│   └── [4.0K]  HT2000toInf
│       └── [4.0K]  0000
│           ├── [ 68M]  hepcnn_10.h5
│           ├── [ 68M]  hepcnn_11.h5
│           ├── [ 68M]  hepcnn_12.h5
│           ├── [ 68M]  hepcnn_13.h5
│           ├── [ 68M]  hepcnn_14.h5
│           ├── [ 68M]  hepcnn_15.h5
│           ├── [ 68M]  hepcnn_16.h5
│           ├── [ 68M]  hepcnn_17.h5
│           ├── [ 68M]  hepcnn_18.h5
│           ├── [ 68M]  hepcnn_19.h5
│           ├── [ 68M]  hepcnn_1.h5
│           ├── [ 68M]  hepcnn_2.h5
│           ├── [ 68M]  hepcnn_3.h5
│           ├── [ 68M]  hepcnn_4.h5
│           ├── [ 68M]  hepcnn_5.h5
│           ├── [ 68M]  hepcnn_6.h5
│           ├── [ 68M]  hepcnn_7.h5
│           ├── [ 68M]  hepcnn_8.h5
│           └── [ 68M]  hepcnn_9.h5
└── [4.0K]  RPV
    └── [4.0K]  Gluino1400GeV
        └── [4.0K]  0000
            ├── [ 31M]  hepcnn_10.h5
            ├── [ 30M]  hepcnn_11.h5
            ├── [ 31M]  hepcnn_12.h5
            ├── [ 30M]  hepcnn_13.h5
            ├── [ 31M]  hepcnn_14.h5
            ├── [ 31M]  hepcnn_15.h5
            ├── [ 31M]  hepcnn_16.h5
            ├── [ 30M]  hepcnn_17.h5
            ├── [ 31M]  hepcnn_18.h5
            ├── [ 31M]  hepcnn_19.h5
            ├── [ 31M]  hepcnn_1.h5
            ├── [ 31M]  hepcnn_20.h5
            ├── [ 31M]  hepcnn_21.h5
            ├── [ 30M]  hepcnn_22.h5
            ├── [ 30M]  hepcnn_23.h5
            ├── [ 30M]  hepcnn_24.h5
            ├── [ 31M]  hepcnn_25.h5
            ├── [ 30M]  hepcnn_26.h5
            ├── [ 31M]  hepcnn_27.h5
            ├── [ 30M]  hepcnn_28.h5
            ├── [ 31M]  hepcnn_29.h5
            ├── [ 30M]  hepcnn_2.h5
            ├── [ 31M]  hepcnn_30.h5
            ├── [ 30M]  hepcnn_31.h5
            ├── [ 30M]  hepcnn_32.h5
            ├── [ 30M]  hepcnn_33.h5
            ├── [ 30M]  hepcnn_34.h5
            ├── [ 31M]  hepcnn_35.h5
            ├── [ 31M]  hepcnn_36.h5
            ├── [ 31M]  hepcnn_37.h5
            ├── [ 31M]  hepcnn_38.h5
            ├── [ 30M]  hepcnn_39.h5
            ├── [ 31M]  hepcnn_3.h5
            ├── [ 31M]  hepcnn_40.h5
            ├── [ 31M]  hepcnn_41.h5
            ├── [ 30M]  hepcnn_42.h5
            ├── [ 31M]  hepcnn_43.h5
            ├── [ 31M]  hepcnn_44.h5
            ├── [ 30M]  hepcnn_45.h5
            ├── [ 31M]  hepcnn_46.h5
            ├── [ 31M]  hepcnn_47.h5
            ├── [ 31M]  hepcnn_48.h5
            ├── [ 31M]  hepcnn_49.h5
            ├── [ 30M]  hepcnn_4.h5
            ├── [ 30M]  hepcnn_5.h5
            ├── [ 31M]  hepcnn_6.h5
            ├── [ 31M]  hepcnn_7.h5
            ├── [ 31M]  hepcnn_8.h5
            └── [ 30M]  hepcnn_9.h5

8 directories, 87 files
```

**Comment by jhgoh** [comment]
training loss 커브 등. 큰 문제는 없는 것으로 보입니다.
image

**Comment by jhgoh** [comment]
알수없는 이유로 현재 multiprocessing시 각 process가 종료되지 않고 계속 running으로 나타나는 현상이 있음.

**Comment by jhgoh** [comment]
file splitting을 바꿔 각 파일의 크기가 작아졌을 때는 문제가 없는 것이 발견됨.

**Comment by jhgoh** [comment]
image

full data활용 performance plot

**Comment by jhgoh** [comment]
```
$ du -sh ../data/CMS2018_unmerged/hdf5_noPU/*/*
29M	../data/CMS2018_unmerged/hdf5_noPU/QCDBkg/HT1000to1500
346M	../data/CMS2018_unmerged/hdf5_noPU/QCDBkg/HT1500to2000
365M	../data/CMS2018_unmerged/hdf5_noPU/QCDBkg/HT2000toInf
569M	../data/CMS2018_unmerged/hdf5_noPU/RPV/Gluino1400GeV
```
$ du -sh ../data/CMS2018_unmerged/hdf5_32PU/*/*
170M	../data/CMS2018_unmerged/hdf5_32PU/QCDBkg/HT1000to1500
645M	../data/CMS2018_unmerged/hdf5_32PU/QCDBkg/HT1500to2000
1.4G	../data/CMS2018_unmerged/hdf5_32PU/QCDBkg/HT2000toInf
1.5G	../data/CMS2018_unmerged/hdf5_32PU/RPV/Gluino1400GeV
```

```
